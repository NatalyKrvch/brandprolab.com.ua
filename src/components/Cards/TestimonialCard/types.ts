export interface TestimonialCardProps {
  text: string;
  amountOfWordsToDisplay: number;
  clientName?: string;
  clientPhotoUrl?: string;
  clientLink?: string;
  className?: string;
}
