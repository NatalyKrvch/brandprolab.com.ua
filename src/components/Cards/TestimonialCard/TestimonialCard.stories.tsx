import type { Meta, StoryObj } from '@storybook/react';

import TestimonialCard from './TestimonialCard';

const meta: Meta<typeof TestimonialCard> = {
  title: 'Components/TestimonialCard',
  component: TestimonialCard,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    text: {
      control: 'text',
      description: 'The testimonial text content.',
      table: {
        type: { summary: 'string' },
      },
    },
    clientName: {
      control: 'text',
      description: 'The name of the client giving the testimonial.',
      table: {
        type: { summary: 'string' },
      },
    },
    clientPhotoUrl: {
      control: 'text',
      description: 'The URL of the client’s profile image.',
      table: {
        type: { summary: 'string' },
      },
    },
    clientLink: {
      control: 'text',
      description:
        'A link to the client’s profile. If empty, the badge is not clickable.',
      table: {
        type: { summary: 'string' },
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof TestimonialCard>;

export const WithClientLink: Story = {
  args: {
    text: 'This is an amazing service! Highly recommend it to everyone.',
    clientName: 'Jane Doe',
    clientPhotoUrl:
      'https://media.vanityfair.com/photos/659d6933b6da4325190e2751/1:1/w_1333,h_1333,c_limit/Lindsay-Lohan.jpg',
    clientLink: 'https://example.com/profile/jane',
  },
};

export const NoClientInfo: Story = {
  args: {
    text: 'Great product! It exceeded my expectations.',
    clientName: '',
    clientPhotoUrl: '',
    clientLink: '',
  },
};
