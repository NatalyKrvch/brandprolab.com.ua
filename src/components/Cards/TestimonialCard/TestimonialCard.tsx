import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';

import { SimpleCard, UserBadge } from '@/components';
import { TESTIMONIAL_TEXT_TEST_ID } from '@/lib/testIDs';
import { getReadMoreText } from '@/utils';

import { TestimonialCardProps } from './types';

const TestimonialCard = ({
  text,
  clientName,
  clientPhotoUrl,
  clientLink,
}: TestimonialCardProps) => {
  const shouldRenderUserBadge = clientName && clientPhotoUrl;

  return (
    <SimpleCard
      borderColor="border-teal-dark"
      className={twMerge(clsx('flex flex-col gap-6', 'p-6 tablet:p-8'))}
    >
      <div
        className={twMerge(
          clsx(
            'text-xl font-normal leading-6 text-black',
            'desktop:text-22 desktop:leading-26',
          ),
        )}
        data-testid={TESTIMONIAL_TEXT_TEST_ID}
      >
        {getReadMoreText(text)}
      </div>
      {shouldRenderUserBadge && (
        <UserBadge
          userName={clientName}
          userPhotoUrl={clientPhotoUrl}
          userLink={clientLink}
        />
      )}
    </SimpleCard>
  );
};

export default TestimonialCard;
