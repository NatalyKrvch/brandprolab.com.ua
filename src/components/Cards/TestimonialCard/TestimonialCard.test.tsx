import '@testing-library/jest-dom';

import { render, screen } from '@testing-library/react';

import {
  SIMPLE_CARD_TEST_ID,
  TESTIMONIAL_TEXT_TEST_ID,
  USER_BADGE_TEST_ID,
} from '@/lib/testIDs';

import TestimonialCard from './TestimonialCard';

jest.mock('@/components/UserBadge', () => ({
  UserBadge: ({
    userName,
    userPhotoUrl,
    userLink,
  }: {
    userName: string;
    userPhotoUrl: string;
    userLink?: string;
  }) => (
    <div data-testid={USER_BADGE_TEST_ID}>
      <span>{userName}</span>
      <img src={userPhotoUrl} alt={userName} />
      {userLink && <a href={userLink}>Profile</a>}
    </div>
  ),
}));

describe('TestimonialCard component', () => {
  const defaultProps = {
    text: 'This is a testimonial text.',
    clientName: 'John Doe',
    clientPhotoUrl: '/john-doe.jpg',
    clientLink: 'https://example.com',
  };

  it('renders the TestimonialCard component', () => {
    const { asFragment } = render(<TestimonialCard {...defaultProps} />);
    const testimonialCard = screen.getByTestId(SIMPLE_CARD_TEST_ID);

    expect(testimonialCard).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the testimonial text correctly', () => {
    render(<TestimonialCard {...defaultProps} />);
    const textElement = screen.getByTestId(TESTIMONIAL_TEXT_TEST_ID);

    expect(textElement).toBeInTheDocument();
    expect(textElement).toHaveTextContent('This is a testimonial text.');
  });

  it('renders the "Читати більше" button', () => {
    render(<TestimonialCard {...defaultProps} />);
    const readMoreButton = screen.getByRole('button', {
      name: /читати більше/i,
    });

    expect(readMoreButton).toBeInTheDocument();
  });

  it('renders UserBadge when clientName and clientPhotoUrl are provided', () => {
    const { asFragment } = render(<TestimonialCard {...defaultProps} />);
    const userBadge = screen.getByTestId(USER_BADGE_TEST_ID);

    expect(userBadge).toBeInTheDocument();
    expect(userBadge).toHaveTextContent('John Doe');
    expect(userBadge.querySelector('img')).toHaveAttribute(
      'src',
      '/john-doe.jpg',
    );
    expect(userBadge.querySelector('a')).toHaveAttribute(
      'href',
      'https://example.com',
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render UserBadge if clientName or clientPhotoUrl is missing', () => {
    render(<TestimonialCard text="Test text" />);
    expect(screen.queryByTestId(USER_BADGE_TEST_ID)).not.toBeInTheDocument();
  });
});
