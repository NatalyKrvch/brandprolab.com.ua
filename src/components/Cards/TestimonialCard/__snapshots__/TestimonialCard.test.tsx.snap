// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestimonialCard component renders UserBadge when clientName and clientPhotoUrl are provided 1`] = `
<DocumentFragment>
  <div
    class="bg-white flex flex-col justify-between p-6 tablet:p-8 undefined border border-teal-dark rounded-24 tablet:rounded-32 desktop:rounded-40 w-full bg-cover bg-center"
    data-testid="simple-card-test-id"
  >
    <div
      class="text-xl font-normal leading-6 text-black desktop:text-22 desktop:leading-26"
      data-testid="testimonials-text-test-id"
    >
      This is a testimonial text.
    </div>
    <div
      data-testid="user-badge-test-id"
    >
      <span>
        John Doe
      </span>
      <img
        alt="John Doe"
        src="/john-doe.jpg"
      />
      <a
        href="https://example.com"
      >
        Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TestimonialCard component renders the TestimonialCard component 1`] = `
<DocumentFragment>
  <div
    class="bg-white flex flex-col justify-between p-6 tablet:p-8 undefined border border-teal-dark rounded-24 tablet:rounded-32 desktop:rounded-40 w-full bg-cover bg-center"
    data-testid="simple-card-test-id"
  >
    <div
      class="text-xl font-normal leading-6 text-black desktop:text-22 desktop:leading-26"
      data-testid="testimonials-text-test-id"
    >
      This is a testimonial text.
    </div>
    <div
      data-testid="user-badge-test-id"
    >
      <span>
        John Doe
      </span>
      <img
        alt="John Doe"
        src="/john-doe.jpg"
      />
      <a
        href="https://example.com"
      >
        Profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;
